pylint/checkers/non_ascii_names.py

```diff
@@ -0,0 +1,237 @@
+            "non-ascii-name",
+            (
+                "Used when the name contains at least one non-ASCII unicode character. "
+                "See https://www.python.org/dev/peps/pep-0672/#confusable-characters-in-identifiers"
+                " for a background why this could be bad."
+            ),
+        ),
+    }
+
+    name = "NonASCII-Checker"
+
+    def __init__(self, linter):
+        super().__init__(linter)
+        self._non_ascii_rgx_compiled = re.compile("[^A-Za-z0-9_]")
+
+    def _raise_name_warning(
+        self,
+        node: nodes.NodeNG,
+        node_type: str,
+        name: str,
+    ) -> None:
+        type_label = HUMAN_READABLE_TYPES[node_type]
+        args = (type_label.capitalize(), name)
+
+        self.add_message(
+            "non-ascii-name", node=node, args=args, confidence=interfaces.HIGH
+        )
+        self.linter.stats.increase_bad_name(node_type, 1)
+
+    def _check_name(
+        self,
+        node_type: str,
+        names: Union[str, List[str]],
+        node: Union[nodes.NodeNG, _AsciiOnlyCheckedNode],
+    ) -> None:
+        """check for a name if it's using non ASCII and set message if it does
+
+        Also set the dynamic attribute ``_is_ascii_only`` that is used to
+        determine if a node has been already checked, so we don't have to handle
+        too many edge cases.
+        """
+        if not isinstance(names, list):
+            names = [names]
+
+        current_state = getattr(node, "_is_ascii_only", True)
+        for name in names:
+            if name is None:
+                # For some nodes i.e. *kwargs from a dict, the name will be empty
+                continue
+            non_ascii_match = self._non_ascii_rgx_compiled.match(name)
+            if not (Py37Str(name).isascii() and non_ascii_match is None):
+                # Note that we require the `.isascii` method as it is fast and
+                # handles the complexities of unicode, so we can use simple regex.
+                self._raise_name_warning(node, node_type, name)
+                current_state = False
+
+        node._is_ascii_only = current_state  # pylint: disable=protected-access
+
+    def _recursive_check_names(self, args: Iterable[nodes.AssignName]):
+        """check names in a possibly recursive list <arg>"""
+        for arg in args:
+            if isinstance(arg, nodes.AssignName):
+                self._check_name("argument", arg.name, arg)
+            else:
+                # pylint: disable-next=fixme
+                # TODO: Check if we can remove this branch because of
+                #       the up to date astroid version used
+                self._recursive_check_names(arg.elts)
+
+    @pylint.checkers.utils.check_messages("non-ascii-name")
+    def visit_module(self, node: nodes.Module) -> None:
+        # pylint: disable-next=fixme
+        # TODO check how the node name is defined when calling the real checker
```

Comment:
Is this a TODO to be handled in this PR?

Reply:
I change this in the new PR. Added a functional test for it and do understand it now (didn't before).